scipy/linalg/_linsolve.pyx.in

```diff
@@ -0,0 +1,603 @@
+# precomputed version of the numpy.can_cast() table.
+
+# output of np.typecodes['All']
+cdef str casting_key = '?bhilqpBHILQPefdgFDGSUVOMm'
+
+#                          f  d  F  D
+casting_table = np.array([[1, 1, 1, 1],  # ?
+                          [1, 1, 1, 1],  # b
+                          [1, 1, 1, 1],  # h
+                          [0, 1, 0, 1],  # i
+                          [0, 1, 0, 1],  # l
+                          [0, 1, 0, 1],  # q
+                          [0, 1, 0, 1],  # p
+                          [1, 1, 1, 1],  # B
+                          [1, 1, 1, 1],  # H
+                          [0, 1, 0, 1],  # I
+                          [0, 1, 0, 1],  # L
+                          [0, 1, 0, 1],  # Q
+                          [0, 1, 0, 1],  # P
+                          [1, 1, 1, 1],  # e
+                          [1, 1, 1, 1],  # f
+                          [0, 1, 0, 1],  # d
+                          [0, 1, 0, 1],  # g
+                          [0, 0, 1, 1],  # F
+                          [0, 0, 0, 1],  # D
+                          [0, 0, 0, 1],  # G
+                          [0, 0, 0, 0],  # S
+                          [0, 0, 0, 0],  # U
+                          [0, 0, 0, 0],  # V
+                          [0, 0, 0, 0],  # O
+                          [0, 0, 0, 0],  # M
+                          [0, 0, 0, 0]],  # m
+                         dtype=bool)
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.nonecheck(False)
+cpdef int get_lapack_flavor(str a, str b, str K = casting_key,
+                            const cnp.npy_bool[:, ::1]C = casting_table):
+    cdef int ind
+    cdef size_t rowa = K.find(a)
+    cdef size_t rowb = K.find(b)
+    if rowa == -1 or rowb == -1:
+        raise TypeError('Data type is not known to NumPy.')
+
+    for ind in range(4):
+        if C[rowa, ind] and C[rowb, ind]:
+            return ind
+
+    return -1
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.nonecheck(False)
+cpdef bint array_is_finite(lapack_t[:, ::1]A) nogil:
+    """
+    Runs over the array and checks for infs and nans. Compared
+    to asarray_chkfinite the performance gets better as the
+    array size increases. Starting from n=3 it is ~x2 faster
+    for finite arrays and faster if array has inf/nan due to
+    early exit capability.
+    """
+    cdef int n = <int>A.shape[0]
+    cdef int r, c
+    cdef lapack_t entry
+    if lapack_t in lapack_cz_t:
+        for r in xrange(n):
+            for c in xrange(n):
+                entry = A[r, c]
+                if not(isfinite(entry.real) and isfinite(entry.imag)):
+                    return False
+    else:
+        for r in xrange(n):
+            for c in xrange(n):
+                if (not isfinite(A[r, c])):
+                    return False
+    return True
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef inline bint _ones_on_diagonal(lapack_t[:, ::1] a) nogil:
```

Comment:
`a` can probably be `const` here (and other places)?

Reply:
I agree but that's an open bug https://github.com/cython/cython/issues/3783

Reply:
In fact I'm starting to feel like tempita is definitely way better than this fused type stuff and then we won't have that bug biting us.

Reply:
Okay, I saw https://github.com/cython/cython/issues/3222 but didn't realize it wasn't backported.  I think (fingers crossed) that a smart compiler would do the right thing here (i.e. notice that this should be const), so nothing much more to do here if the Cython bug is still around